### DISCUSSION TITLE
PlanetScale connection issues with Vercel deployment

### DISCUSSION THREAD

**Discussion by jbarrasso** [prisma/prisma#13738] [Q&A]
Hello,

I had a working deployed Vercel application that could read/write to my pscale database with no issues.

It was only until yesterday when I started deleting previous preview deployments that my connection issues started happening... on my production app I can write sometimes but other times I can't... and it appears to be at random.

`[PUT] /api/users 11:48:49:30 2022-06-09T18:48:49.441Z bf98c04c-5ec0-4656-833c-335bed08912d ERROR Unhandled Promise Rejection {"errorType":"Runtime.UnhandledPromiseRejection","errorMessage":"Error: Can't reach database server at pitqyt0m4npt.us-east-4.psdb.cloud:3306\n\nPlease make sure your database server is running at pitqyt0m4npt.us-east-4.psdb.cloud:3306.","reason":{"errorType":"Error","errorMessage":"Can't reach database server at pitqyt0m4npt.us-east-4.psdb.cloud:3306\n\nPlease make sure your database server is running at pitqyt0m4npt.us-east-4.psdb.cloud:3306.","code":"P1001","clientVersion":"3.14.0","meta":{"database_host":"pitqyt0m4npt.us-east-4.psdb.cloud","database_port":3306},"stack":["Error: Can't reach database server at pitqyt0m4npt.us-east-4.psdb.cloud:3306","","Please make sure your database server is running at pitqyt0m4npt.us-east-4.psdb.cloud:3306."," at Object.request (/var/task/node_modules/@prisma/client/runtime/index.js:45405:15)"," at async PrismaClient._request (/var/task/node_modules/@prisma/client/runtime/index.js:46301:18)"," at async updateOverallStats (/var/task/.next/server/pages/api/toadStats.js:259:17)"]},"promise":{},"stack":["Runtime.UnhandledPromiseRejection: Error: Can't reach database server at pitqyt0m4npt.us-east-4.psdb.cloud:3306","","Please make sure your database server is running at pitqyt0m4npt.us-east-4.psdb.cloud:3306."," at process.<anonymous> (file:///var/runtime/index.mjs:775:15)"," at process.emit (node:events:539:35)"," at emit (node:internal/process/promises:140:20)"," at processPromiseRejections (node:internal/process/promises:274:27)"," at processTicksAndRejections (node:internal/process/task_queues:97:32)"]} Unknown application error occurred`

Things to note: 

- My app works perfectly fine in a local environment. I can read/write with no issues.
- I removed the Planetscale/Vercel integration and re-added it to generate new connection strings multiple times with no luck
- I tried this with a new Planetscale database with no luck
- I tried this with a new Vercel project under a Pro plan with no luck
- I did speak with Planetscale support who looked into the logs of the database in question. They showed me what looked to be a timeout on the transactions I am performing.

![image](https://user-images.githubusercontent.com/18688696/172962328-7988cff9-228b-4893-81b0-d6dde913e94d.png)

My question is: how is that my app works perfectly fine locally but not deployed? It's something specifically about the Vercel deployment that is causing issues. What else could be the cause? Any suggestions? Thanks

**Comment by nurul3101** [comment]
Hey @jbarrasso 👋 

Are you using [interactiveTransactions](https://www.prisma.io/docs/concepts/components/prisma-client/transactions#interactive-transactions-in-preview) or the normal [$transaction](https://www.prisma.io/docs/concepts/components/prisma-client/transactions#the-transaction-api) API?

From the logs as you said it seems you are running into 20 seconds timeout limit from PlanetScale.

If you are using interactiveTransactions then it might be helpful to increase the timeout value by configuring the second parameter of API like this:

```javascript
await prisma.$transaction(
  async (prisma) => {
    // Code running in a transaction...
  },
  {
    maxWait: 5000, // default: 2000
    timeout: 10000, // default: 5000
  }
)
```
Reference for interactiveTransaction: [Link](https://www.prisma.io/docs/concepts/components/prisma-client/transactions#interactive-transactions-in-preview)

**Reply by jbarrasso** [reply]
Thanks for the reply @nurul3101.

I am using the interactiveTransactions preview feature because I need to perform some game logic in between writes.

I tried increasing the maxWaits and timeout to 5000 and 10000, then even 20000 and 25000 but it didn't make a difference.

What's unusual is that my queries are not even that complex, it's only a few writes.. see below


```
            const eat = async() => {
                let fedValue: number
                let energyValue: number
                let happinessValue: number
                let healthValue: number
        
                if (canFeed()) {
                    await prisma.$transaction(async (prisma) => {
                         prisma.toadz.update({
                            where: { toadId : selectedToad[0].toadId },
                            data: { fed : selectedToad[0].fed + 3,
                                    happiness: selectedToad[0].happiness + 1,
                                    health: selectedToad[0].health + 1 }
                        })
                        fedValue = selectedToad[0].fed + 3
                        energyValue = selectedToad[0].energy
                        happinessValue = selectedToad[0].happiness + 1
                        healthValue = selectedToad[0].health + 1
                        
                        updateOverallStats(fedValue, energyValue, happinessValue, healthValue)
                        grantXp(40)
                    },
                    {
                        maxWait: 20000, // default: 2000
                        timeout: 25000, // default: 5000
                    })
                    res.status(200).json(
                        {
                            message: `You fed toad some pizza... Delicious!`,
                            animation: 'pizza',
                            points: thisOwner.points
                        }
                    )
                }
            }
```

updateOverallStats

```
            async function updateOverallStats(fedValue: number, energyValue: number, happinessValue: number, healthValue: number) {
                if (fedValue < 0) {
                    await prisma.toadz.update({
                        where: { toadId : selectedToad[0].toadId },
                        data: { fed : 0}
                    })
                    fedValue = 0
                } else if (fedValue > 10) {
                    await prisma.toadz.update({
                        where: { toadId : selectedToad[0].toadId },
                        data: { fed : 10}
                    })
                    fedValue = 10
                }
                if (energyValue < 0) {
                    await prisma.toadz.update({
                        where: { toadId : selectedToad[0].toadId },
                        data: { energy : 0}
                    })
                    energyValue = 0
                } else if (energyValue > 10) {
                    await prisma.toadz.update({
                        where: { toadId : selectedToad[0].toadId },
                        data: { energy : 10}
                    })
                    energyValue = 10
                }
                if (happinessValue < 0) {
                    await prisma.toadz.update({
                        where: { toadId : selectedToad[0].toadId },
                        data: { happiness : 0}
                    })
                    happinessValue = 0
                } else if (happinessValue > 10) {
                    await prisma.toadz.update({
                        where: { toadId : selectedToad[0].toadId },
                        data: { happiness : 10}
                    })
                    happinessValue = 10
                }
                if (healthValue < 0) {
                    await prisma.toadz.update({
                        where: { toadId : selectedToad[0].toadId },
                        data: { health : 0}
                    })
                    healthValue = 0
                } else if (healthValue > 10) {
                    await prisma.toadz.update({
                        where: { toadId : selectedToad[0].toadId },
                        data: { health : 10}
                    })
                    healthValue = 10
                }
        
                const newOverall = Math.round(((fedValue + energyValue + happinessValue + healthValue) / 4))
                await prisma.toadz.update({
                    where: { toadId : selectedToad[0].toadId },
                    data: { overall: newOverall}
                })
            }
```

grantXp

```
            async function grantXp(actionXp: number) {
                //add exponential feature
                let currentXp = selectedToad[0].xp
        
                if ((currentXp + actionXp) >= 100) {
                    let leftoverXp = (currentXp + actionXp) - 100
                    await prisma.toadz.update({
                        where: { toadId : selectedToad[0].toadId },
                        data: { xp : leftoverXp,
                                level: selectedToad[0].level + 1 }
                    })
                    await prisma.user.update({
                        where: { address: selectedToad[0].userId },
                        data: { points: {increment: 100}}
                    })
                } else {
                    let newXp = currentXp + actionXp
                    await prisma.toadz.update({
                        where: { toadId : selectedToad[0].toadId },
                        data: { xp : newXp}
                    })
                    await prisma.user.update({
                        where: { address: selectedToad[0].userId },
                        data: { points: {increment: 10}}
                    })
                }
            }
```

As you can see, the amount of writes is only between 4 and 8 depending on what the current game values are. How is that these writes are taking so much time? The amount of total IDs in the toadz table is 7,000 and there are only about 3 users so far.

cc @janpio

**Comment by janpio** [comment]
I replied in another discussion that included similar information: https://github.com/prisma/prisma/discussions/13595#discussioncomment-2923423

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
